lib/cn/cn.c

```diff
@@ -933,58 +935,65 @@ cn_open(
-     */
-    {
-        struct ekey_generator *egen;
+    err = cndb_cn_instantiate(cndb, cnid, &ctx, (void *)cn_kvset_cb);
+    if (ev(err))
+        goto err_exit;
 
-        egen = ekgen_create(kvs_name, cn->cp);
-        assert(egen);
+    if (cn->cn_tree->ct_route_map) {
+        char kbuf[HSE_KVS_KEY_LEN_MAX];
+        struct cn_tree_node *tn, *next;
+        struct route_node *rn;
+        uint klen;
 
-        for (uint i = 0; i < cn->cp->fanout; i++) {
-            struct cn_tree_node *tn;
+        list_for_each_entry_reverse_safe(tn, next, &cn->cn_tree->ct_nodes, tn_link) {
+            if (tn->tn_nodeid > 0) {
+                cn_tree_node_get_max_key(tn, kbuf, sizeof(kbuf), &klen);
 
-            tn = cn_node_alloc(cn->cn_tree, i + 1);
-            if (!tn) {
-                err = merr(ENOMEM);
-                break;
+                log_info("nodeid %lu, klen %u", tn->tn_nodeid, klen);
+
+                tn->tn_route_node = route_map_insert(cn->cn_tree->ct_route_map, tn, kbuf, klen);
+                if (!tn->tn_route_node) {
+                    err = merr(EINVAL);
+                    goto err_exit;
+                }
             }
+        }
 
-            map_insert_ptr(ctx.nodemap, tn->tn_nodeid, tn);
+        klen = sizeof(kbuf);
+        memset(kbuf, -1, klen);
 
-            list_add_tail(&tn->tn_link, &cn->cn_tree->ct_nodes);
+        rn = route_map_lookup(cn->cn_tree->ct_route_map, kbuf, klen);
+        if (rn) {
+            tn = route_node_tnode(rn);
+            if (!tn)
+                abort();
 
-            if (cn->cn_tree->ct_route_map) {
-                char ekbuf[HSE_KVS_KEY_LEN_MAX];
-                size_t eklen;
+            log_info("updating node %lu right edge key", tn->tn_nodeid);
 
-                eklen = ekgen_generate(egen, ekbuf, sizeof(ekbuf), i);
-                if (eklen != 0) {
-                    tn->tn_route_node = route_map_insert(cn->cn_tree->ct_route_map, tn,
-                        ekbuf, eklen);
-                }
+            err = route_node_key_modify(cn->cn_tree->ct_route_map, rn, kbuf, klen);
+            if (err)
+                goto err_exit;
+        } else {
+            tn = cn_node_alloc(cn->cn_tree, cndb_nodeid_mint(cndb));
+            if (!tn) {
+                err = merr(ENOMEM);
+                goto err_exit;
             }
-        }
-
-        ekgen_destroy(egen);
-    }
 
-    if (ev(err))
-        goto err_exit;
+            log_info("creating node %lu right edge key", tn->tn_nodeid);
 
-    cn_tree_set_initial_dgen(cn->cn_tree, 0);
-    err = cndb_cn_instantiate(cndb, cnid, &ctx, (void *)cn_kvset_cb);
-    if (ev(err))
-        goto err_exit;
+            list_add_tail(&tn->tn_link, &cn->cn_tree->ct_nodes);
+            tn->tn_route_node = route_map_insert(cn->cn_tree->ct_route_map, tn, kbuf, klen);
+        }
+    }
 
     map_destroy(ctx.nodemap);
     ctx.nodemap = 0;
```

Comment:
Can you move this to right after the call to `cndb_cn_instantiate()` (before checking for errors).  Then you can also remove it from  `err_exit`.

Reply:
Lemme look, not sure that makes sense...

Reply:
Alex, have a look at the second commit, I'm not understanding what you're asking here...

Reply:
See other comment about hiding use of `ctx` from this function.